cloudinit/cmd/main.py

```diff
@@ -334,6 +337,90 @@ def _should_bring_up_interfaces(init, args):
     return not args.local
 
 
+def _should_wait_via_cloud_config(
+    raw_config: Optional[Union[str, bytes]]
+) -> Tuple[bool, Reason]:
+    if not raw_config:
+        return False, "no configuration found"
+
+    # If our header is anything other than #cloud-config, wait
+    possible_header: Union[bytes, str] = raw_config.strip()[:13]
+    if isinstance(possible_header, str):
+        decoded_header = possible_header
+    elif isinstance(possible_header, bytes):
+        try:
+            decoded_header = possible_header.decode("utf-8")
+        except UnicodeDecodeError:
+            return True, "Binary user data found"
+    if not decoded_header.startswith("#cloud-config"):
+        return True, "non-cloud-config user data found"
+
+    try:
+        parsed_yaml = yaml.safe_load(raw_config)
+    except Exception as e:
+        log_with_downgradable_level(
+            logger=LOG,
+            version="24.4",
+            requested_level=logging.WARNING,
+            msg="Unexpected failure parsing userdata: %s",
+            args=e,
+        )
+        return True, "failed to parse user data as yaml"
+
+    # These all have the potential to require network access, so we should wait
+    if "write_files" in parsed_yaml and any(
+        "source" in item for item in parsed_yaml["write_files"]
+    ):
+        return True, "write_files with source found"
```

Comment:
This could be limited a bit more, since source.uri could be a local file with either `file:` or `/` prefix.
```suggestion
    if "write_files" in parsed_yaml:
        for item in parsed_yaml["write_files"]:
            source_uri = item.get("source", {}).get("uri", "")
            if source_uri and source_uri[0] != "/" and source_uri[:5] != "file:":
                return True, "write_files with source found"
```

Reply:
Thanks, applied